[benchmark] Add some benchmarks for String breadcrumbs

lorentey opened the pull request:
This adds some benchmarks for `String` breadcrumbing, via low-level stdlib entry points exposed by `String`.

- `StringBreadcrumbs.UTF16ToIndex`: Convert some random UTF-16 offsets into String indices.
- `StringBreadcrumbs.IndexToUTF16`: Convert some random String indices into UTF-16 offsets.
- `StringBreadcrumbs.UTF16ToIndexRange`: Split a string into a bunch of random slices and convert their UTF-16 offsets into String index ranges.
- `StringBreadcrumbs.IndexToUTF16Range`: Do the same in reverse direction.
- `StringBreadcrumbs.CopyUTF16CodeUnits`: Copy UTF-16 code units from random ranges.
- `StringBreadcrumbs.MutatedUTF16ToIndex`: This is like `UTF16ToIndex` but appends to the string after every index conversion. In effect, this tests breadcrumb creation performance.
- `StringBreadcrumbs.MutatedIndexToUTF16`: This is like `IndexToUTF16` but appends to the string after every index conversion. In effect, this tests breadcrumb creation performance.

https://bugs.swift.org/browse/SR-9226

lorentey committed:
[benchmark] Add some benchmarks for String breadcrumbs

lorentey commented:
The CopyUTF16CodeUnits benchmark here will crash until https://github.com/apple/swift/pull/20768 lands.

palimondo commented:
@lorentey #20667 has landed. You should be able to run these now with swift-ci, too.

lorentey commented:
Please test with the following PR: https://github.com/apple/swift/pull/20768
@please benchmark

lorentey commented:
https://github.com/apple/swift/pull/20768
@swift-ci smoke test

lorentey commented:
https://github.com/apple/swift/pull/20768
@swift-ci benchmark

palimondo commented on benchmark/single-source/StringBreadcrumbs.swift:
```diff
@@ -0,0 +1,494 @@
+  }
+
+  @inline(never)
+  override func run() {
```
Maybe make these `final` to have static dispatch?

lorentey replied:
I don't think that's necessary -- these are non-@objc internal classes so the optimizer should be able to see that there aren't any further overrides. (And the benchmark payload will hopefully be overwhelmingly larger than a virtual method call.)

(Lots of thanks for the run_smoke_bench fix, btw! I wasn't aware we were matching benchmark names like that.)

palimondo replied:
Sure, assuming you've verified this really has no effect...

palimondo commented on benchmark/single-source/StringBreadcrumbs.swift:
```diff
@@ -0,0 +1,494 @@
+  MutatedUTF16ToIndex(workload: shortMixedWorkload, count: 50).info,
+  MutatedIndexToUTF16(workload: shortASCIIWorkload, count: 50).info,
+  MutatedIndexToUTF16(workload: shortMixedWorkload, count: 50).info,
+]
```
And these names are over 40… :-/

lorentey replied:
I don't see a way to shorten them without compromising clarity; they're already quite dense.

palimondo replied:
I see two opportunities to shorten the final combined names:

1. `Breadcrumbs` as a family name is unique and memorable enough.
2. The `long` and `short` variants seem to be exclusively divided between the other 4 groups, so these prefixes can be dropped.

With these changes, the longest names would be  is `MutatedUTF16ToIndex`, the final results would be 😎 37 characters long: 

````
✅ Breadcrumbs.MutatedUTF16ToIndex.ASCII
✅ Breadcrumbs.MutatedIndexToUTF16.Mixed
````

lorentey replied:
I'm okay with just Breadcrumbs. `longASCII` and `shortASCII` are quite different workloads though. 

I held my nose and replaced `Index` with `Idx`, `shortASCII` with `ASCII` and `shortMixed` with `Mixed`, leaving the `long`* variants intact. This leaves the names ugly but acceptable, and it gets them all below 40.

palimondo replied:
> longASCII and shortASCII are quite different workloads though.
How so? I only see the same string repeated different times. That seems to me just like a different way to scale the workload size, instead of setting the inner-loop multiplier. Am I mistaken? Is the difference some kind of small string optimization behind the scenes? If so, this should probably be also documented in the comments.

lorentey commented:
@palimondo The doctor complained about the names and the "very wide range of memory used", but I scaled things so that it likes the timings.

lorentey commented:
@milseman I did some spot checks -- the heavy stack frames were inside the target entry points and looked breadcrumby enough. I wasn't looking for missed optimization opportunities though. 

I'll have a more detailed look tomorrow, just to make sure these are measuring what I think they're measuring.

palimondo commented:
> The doctor complained about the names and the "very wide range of memory used"
Interesting. Could you re-check those with `--verbose` flag, to see how bad it is?
Tip: You can also run just select benchmarks by their test numbers, which can come in handy if you want to focus on sub-ranges, like this:
````
$ ./Benchmark_Driver run -i 3 --output-dir log {204..217} {670..683}
````

lorentey commented:
@palimondo Here's a quick sample of the memory-related output:

```
memory: 'StringBreadcrumbs.CopyUTF16CodeUnits.shortASCII' has very wide range of memory used between independent, repeated measurements.
memory: StringBreadcrumbs.CopyUTF16CodeUnits.shortASCII mem_pages [i1, i2]:  min=[42, 46] 𝚫=4 R=[4, 17]
memory: 'StringBreadcrumbs.IndexToUTF16.longASCII' has very wide range of memory used between independent, repeated measurements.
memory: StringBreadcrumbs.IndexToUTF16.longASCII mem_pages [i1, i2]:  min=[1108, 1119] 𝚫=11 R=[44, 31]
memory: 'StringBreadcrumbs.IndexToUTF16.longMixed' has very wide range of memory used between independent, repeated measurements.
memory: StringBreadcrumbs.IndexToUTF16.longMixed mem_pages [i1, i2]:  min=[323, 326] 𝚫=3 R=[36, 36]
memory: 'StringBreadcrumbs.IndexToUTF16Range.longASCII' has very wide range of memory used between independent, repeated measurements.
memory: StringBreadcrumbs.IndexToUTF16Range.longASCII mem_pages [i1, i2]:  min=[1116, 1125] 𝚫=9 R=[33, 29]
```

palimondo commented on benchmark/single-source/StringBreadcrumbs.swift:
```diff
@@ -0,0 +1,494 @@
+    👍👩‍👩‍👧‍👧👨‍👨‍👦‍👦🇺🇸🇨🇦🇲🇽👍🏻👍🏼👍🏽👍🏾👍🏿the quick brown fox👍🏿👍🏾👍🏽👍🏼👍🏻🇲🇽🇨🇦🇺🇸👨‍👨‍👦‍👦👩‍👩‍👧‍👧👍
+    siebenhundertsiebenundsiebzigtausendsiebenhundertsiebenundsiebzig
+    今回のアップデートでSwiftに大幅な改良が施され、安定していてしかも直感的に使うことができるAppleプラットフォーム向けプログラミング言語になりました。
+    Worst thing about working on String is that it breaks *everything*. Asserts, debuggers, and *especially* printf-style debugging 😭
```
I hear you… 😂

milseman replied:
https://twitter.com/Ilseman/status/850030173613309953

palimondo commented on benchmark/single-source/StringBreadcrumbs.swift:
```diff
@@ -0,0 +1,494 @@
+//===----------------------------------------------------------------------===//
+
+import TestsUtils
+
```
Could you add a description of what are the "Breadcrumbs" tested here?
I'm still in the dark even though I have read most of https://forums.swift.org/t/string-s-abi-and-utf-8/17676/

palimondo commented on benchmark/single-source/StringBreadcrumbs.swift:
```diff
@@ -0,0 +1,494 @@
+  @inline(never)
+  override func run() {
+    for range in inputIndices {
+      print(range)
```
This line is breaking the benchmark run. From https://ci.swift.org/job/swift-PR-osx-perf/607/console:
````
16:41:04 583,StrToInt,3,1615,1668,1633,30,1616
16:41:04 584,StringAdder,3,461,461,461,0,461
16:41:04 2303..<2325
16:41:04 3435..<3454
16:41:04 4487..<4501
16:41:04 1556..<1580
16:41:04 1124..<1141
````

palimondo replied:
My bad. The crash was due to PR 20768 not being merged at the time it was run.
But still… that debug print needs to go! 😝

lorentey replied:
Oops, d'oh...

milseman commented:
@swift-ci please smoke test os x platform

milseman commented:
@swift-ci please benchmark

palimondo reviewed (changes requested):
This is very fresh approach to writing benchmarks! 👍

So far I have these comments:
* [Documentation block at the beginning](https://github.com/apple/swift/pull/20769/files/68e886bfdbfc48ce19f46b577233895c11557ae7#r236449422) that explains what are the `Breadcrumbs` being tested here, would be very helpful. 
* The test names would ideally fit under 40 characters, [renames suggested here](https://github.com/apple/swift/pull/20769/files/68e886bfdbfc48ce19f46b577233895c11557ae7#r236456089). 
* The [debug print](https://github.com/apple/swift/pull/20769/files/68e886bfdbfc48ce19f46b577233895c11557ae7#r236458985) should be removed.

palimondo commented on benchmark/single-source/StringBreadcrumbs.swift:
```diff
@@ -0,0 +1,494 @@
+
+  @inline(never)
+  override func run() {
+    for range in inputIndices {
```
What about moving this `for-in` inside the `withUnsafeMutableBufferPointer` closure?

palimondo commented on benchmark/single-source/StringBreadcrumbs.swift:
```diff
@@ -0,0 +1,494 @@
+}
+
+
+/// This is like `UTF16ToIndex` but appends to the string after every index
```
Copy-paste typo? Shouldn't this be like `IndexToUTF16`?

milseman commented:
@palimondo where is the 40-character limit coming from and is there any way to get more? Is it Github's in-PR renderer that's constraining this?

palimondo commented:
@milseman 
> where is the 40-character limit coming from and is there any way to get more? Is it Github's in-PR renderer that's constraining this?

Yes, primarily GitHub's fixed table width. As the above report shows, these name at 48 chars are filling the available entire space. But this is a best case scenario, as the `-O` runtimes are under 1000 μs . When you combine this with longer runtimes and all the benchmarks, it triggers horizontal scrolling.

Also formatting the result table in console (`Benchmark_Driver`'s output as run through `build_script`). And using all of benchmark reports on mobile devices, see https://github.com/apple/swift/pull/20334#discussion_r234527138.

palimondo commented:
The variance in `IndexToUTF16` group worries me:


TEST | MIN | MAX | MEAN | MAX_RSS
-- | -- | -- | -- | --
StringBreadcrumbs.IndexToUTF16.longASCII | 660 | 2211 | 1177 | —
StringBreadcrumbs.IndexToUTF16.longMixed | 791 | 1312 | 965 | —
StringBreadcrumbs.IndexToUTF16Range.longASCII | 306 | 1786 | 800 | —
StringBreadcrumbs.IndexToUTF16Range.longMixed | 387 | 390 | 388| —

Only the `longMixed` look fine. This might be background process interference, with `run_smoke_bench`. It still uses averaging… instead of going with `--num-iters=1`.

@lorentey Could you please check locally the variance isn't this crazy, when measured properly?
````
$ ./Benchmark_Driver run -i 3 -f StringBreadcrumbs.IndexToUTF16Range

lorentey committed:
[benchmark] Breadcrumbs: Improve docs; shorten names

lorentey committed:
[benchmark] Remove stray print

lorentey committed:
[benchmark] Eliminate output buffers; collapse run() into run(iterations:); fix memory accumulation

I planned to run some validation on the accumulated output in tearDown, but benchmarks don’t seem the right place for doing that.

Having the run() vs run(iterations:) split interfered with getting e.g. Array.withUnsafeMutableBufferPointer moved out of the outer loop.

MutatedIdxToUTF16 and its dual used to append a space to the input string on every single iteration of the inner loop, accumulating spaces indefinitely, even across the outer iterations. It now works on a per-iteration copy of the input string and alternates between appending/removing a final character.

lorentey commented:
`IdxToUTF16` looks fine to me locally:

```
$ .../Benchmark_Driver run -i 3 -f Breadcrumbs.IdxToUTF16
#,TEST,SAMPLES,MIN(μs),Q1(μs),MEDIAN(μs),Q3(μs),MAX(μs),MAX_RSS(B)
45,Breadcrumbs.IdxToUTF16.longASCII,55,421,433,434,435,452,4575232
46,Breadcrumbs.IdxToUTF16.longMixed,58,555,572,574,577,600,1310720
47,Breadcrumbs.IdxToUTF16Range.longASCII,57,202,209,209,210,214,4575232
48,Breadcrumbs.IdxToUTF16Range.longMixed,58,290,299,299,301,311,1327104

Total performance tests executed: 4
```

lorentey commented:
Rebased, applied review notes, cleaned up some things.

@swift-ci please smoke test

lorentey commented:
@swift-ci please smoke test

lorentey commented:
@swift-ci please smoke test

lorentey commented:
@swift-ci please smoke test

lorentey commented:
@swift-ci benchmark

lorentey commented:
@swift-ci benchmark

lorentey commented:
@swift-ci benchmark

lorentey commented:
@swift-ci benchmark

lorentey commented:
@milseman I profiled all tests looking for anomalies; they seem to work as intended. I did not notice anything left unspecialized, but it seems noteworthy that the `IdxToUTF16` and `IdxToUTF16Range` tests spent about half of their time in BidirectionalCollection's index navigation, and only a third in `getBreadcrumb(forIndex:)`. Some internal heuristic may not work right for these tests.

palimondo commented on benchmark/single-source/Breadcrumbs.swift:
```diff
@@ -113,21 +113,17 @@ class BenchmarkBase {
     return BenchmarkInfo(
       name: self.label,
-      runFunction: self.run(iterations:),
+      runFunction: { self.run(iterations: $0) },
```
Is the closure necessary? The original form, directly referencing the method, worked fine, right?

lorentey replied:
Yeah, there isn't any meaningful difference; I just forgot to revert this chunk before committing. It's a harmless change, though, so I'll leave it in.

palimondo commented on benchmark/single-source/Breadcrumbs.swift:
```diff
@@ -449,11 +436,20 @@ class MutatedUTF16ToIdx: BenchmarkBase {
+        blackHole(string._toUTF16Index(offset))
+        if flag {
+          string.append(" ")
+          flag = false
```
This looks like a perfect place to use `Boolean.toggle()`...

palimondo replied:
How about:
````swift
switch flag {
case true: string.append(" ")
case false: string.removeLast()
}
flag.toggle()

palimondo reviewed (approved):
LGTM. Few cosmetic nits in the comments. Thank's for hearing me out, Karoy!

lorentey committed:
[benchmark] More cleanups

lorentey commented:
Please test with the following PR:

https://github.com/apple/swift/pull/20807
@swift-ci please benchmark

lorentey commented:
Please test with the following PR:

https://github.com/apple/swift/pull/20807
@swift-ci please benchmark

lorentey commented:
@swift-ci please smoke check

lorentey commented:
@swift-ci please smoke check

lorentey commented:
@swift-ci please smoke check

lorentey commented:
@swift-ci please smoke check

lorentey commented:
@swift-ci please smoke check

lorentey commented:
@swift-ci please smoke check

palimondo commented on benchmark/single-source/Breadcrumbs.swift:
```diff
@@ -0,0 +1,499 @@
+}
+
+/// This is like `UTF16ToIdx` but appends to the string after every index
+/// conversion. In effect, this tests breadcrumb creation performance.
```
This comment now doesn't exactly reflect the add/remove mutations going on.

Also, without knowing details of the breadcrumbs implementation, it's not clear to me how this mutation is affecting it… Does it mean that the memoized UTF16 representations are thrown away on each mutation? If so, can you add such explanation to this comment?

palimondo replied:
Damn… I was too slow :-) These are not super important...

palimondo commented:
@lorentey BTW, brace for `BenchmarkDoctor` complaining about names not being in `UpperCamelCase`, since the new naming convention isn't in yet… could you maybe help me finish and merge that in the meantime?

palimondo commented:
@lorentey It looks like it didn't run the benchmarks with the other PR. I guess this only works for `test`? // cc @eeckstein

palimondo commented:
Benchmark | MIN | MAX | MEAN
-- | --: | --: | --: 
Breadcrumbs.IdxToUTF16.longASCII | 586 | 2077 | 1083
Breadcrumbs.IdxToUTF16.longMixed | 714 | 724 | 720
Breadcrumbs.IdxToUTF16Range.longASCII | 289 | 1756 | 778
Breadcrumbs.IdxToUTF16Range.longMixed | 371 | 373 | 372

🧐`longMixed` look great, but `longASCII` is weirding me out… You said it [looked fine locally](https://github.com/apple/swift/pull/20769#issuecomment-442569784), so I don't know how to explain the variance...

milseman commented:
@swift-ci please smoke test

palimondo commented:
@lorentey, you didn't answer my question from [older conversation](https://github.com/apple/swift/pull/20769#discussion_r237246638):
> longASCII and shortASCII are quite different workloads though.

How so? I only see the same string repeated different times. That seems to me just like a different way to scale the workload size, instead of setting the inner-loop multiplier. Am I mistaken? Is the difference about some kind of small string optimization behind the scenes? If so, this should probably be also documented in the comments.

palimondo commented:
> 🧐`longMixed` look great, but `longASCII` is weirding me out… You said it looked fine locally, so I don't know how to explain the variance...

Upon further thought, the variance can be explained this way: `run_smoke_bench` initially gathers only 3 samples per benchmark and that's all we get for the added benchmarks, therefore any benchmarks that have _any_ warmup will manifest itself as this variance. 
@eeckstein: _We should gather more independent samples also for the added benchmarks_.

----

@lorentey @milseman There's something strange going on with these benchmarks and/or the breadcrumbs implementation. 

After Karoy ran the `Benchmark_Driver check` locally he sized workloads and iterations of the benchmarks to sit nicely in the hundreds of μs range, see [`StringBreadcrumbs` benchmark report](https://github.com/apple/swift/pull/20769#issuecomment-441849354) above. Then there were two force-pushes that are lost to history, which probably re-adjusted the workloads of `CopyUTF16CodeUnits`, because in the last benchmark [report above](https://github.com/apple/swift/pull/20769#issuecomment-442596088), their runtime dropped by a factor of 10. (We should IMO revert this by increasing the `count` value back to `5_000`.)

TEST | MIN | MAX | MEAN |
--|--|--|--
Breadcrumbs.CopyUTF16CodeUnits.ASCII | 47 | 48 | 48
Breadcrumbs.CopyUTF16CodeUnits.Mixed | 52 | 55 | 53

Strangely, [as soon as 3rd of December](https://github.com/apple/swift/pull/20965#issuecomment-443785665) some of these start to sporadically appear in the `-O` Performance  changes reported by @swift-ci start with very low runtimes:

TEST | OLD | NEW | DELTA| RATIO
--|--|--|--|--
Breadcrumbs.CopyUTF16CodeUnits.ASCII | 20 | 18 | -10.0% | 1.11x
Breadcrumbs.CopyUTF16CodeUnits.Mixed | 68 | 55 | -19.1% | 1.24x

There were no benchmark reports from @swift-ci in any PR's here on GitHub that would include the improvement as intentional change…  I have noticed that something is strange here during my janitor duty, as even on my measly machine the `MutatedIdxToUTF16.ASCII ` and `MutatedUTF16ToIdx.ASCII` dropped to 15 μs and were picked up by the `BenchmarkDoctor` heuristic that flags runtimes below 20 μs (because than even 1 μs change is above the 5% threshold increasing the chances it shows up in report).

I've looked at it in Instruments: 

ASCII workload's stack trace:
````
   2 Benchmark_O 289.0  MutatedIdxToUTF16.run(iterations:)
   1 libswiftCore.dylib 35.0  StringProtocol._toUTF16Offset(_:)
   0 libswiftCore.dylib 12.0  specialized String.UTF16View._nativeGetOffset(for:)
````

Mixed workload's stack trace:
````
  10 Benchmark_O 16815.0  MutatedIdxToUTF16.run(iterations:)
   9 libswiftCore.dylib 15682.0  StringProtocol._toUTF16Offset(_:)
   8 libswiftCore.dylib 15660.0  specialized String.UTF16View._nativeGetOffset(for:)
   7 libswiftCore.dylib 15342.0  specialized _StringGuts.populateBreadcrumbs(_:)
   6 libswiftCore.dylib 15012.0  specialized _StringBreadcrumbs.init(_:)
````

It looks like ASCII workload is not using any breadcrumbs. Is that normal or a bug?!

milseman commented:
ASCII <-> UTF-16 can be calculated in constant time. It shouldn't use breadcrumbs. We can either have a benchmark demonstrating the triviality of this workload, or maybe a SIL-level test as I think the ASCII check is in inlinable code.

palimondo commented:
I don't understand why the performance demonstrated in the ASCII benchmarks changed between when @lorentey wrote them as seen in the [last benchmark report above](https://github.com/apple/swift/pull/20769#issuecomment-442596088) and later, given that [StringBreadcrumbs.swift](https://github.com/apple/swift/commits/master/stdlib/public/core/StringBreadcrumbs.swift) wasn't modified during that time. And as I said before, there are no publicly visible benchmark reports that would show this improvement was intentional effect of some PR. 

Can you please check in your internal performance tracking when these changed? Let's say the `MutatedIdxToUTF16.ASCII ` and `MutatedUTF16ToIdx.ASCII` in particular... cc @eeckstein

milseman commented:
https://github.com/apple/swift/pull/20848

palimondo commented:
Oh! I have apparently searched through pulls without the proper filters... Thank you!

palimondo commented:
> We can either have a benchmark demonstrating the triviality of this workload, or maybe a SIL-level test as I think the ASCII check is in inlinable code.

I guess this is fine as is — I'll keep an eye out to see if the sub 20 μs runtime will make these benchmarks show up as false changes. I'm sorry to have bothered you with this. And thanks again for digging up the relevant PR, @milseman!
